File: Sheets/src/neetcode/practice/Intervals/_3_nonOverlappingIntervals.java
package neetcode.practice.Intervals;

import java.util.Arrays;

public class _3_nonOverlappingIntervals {
  public int eraseOverlapIntervals(int[][] intervals) {
    if (intervals.length == 1) return 0;

    Arrays.sort(intervals, (a, b) -> Integer.compare(a[0], b[0]));
    int cnt = 0;
    int prevEnd = intervals[0][1];

    for (int i = 1; i < intervals.length; i++) {
      int start = intervals[i][0];
      int end = intervals[i][1];
      if (start >= prevEnd) {
        prevEnd = end;
      } else {
        cnt++;
        prevEnd = Math.min(end, prevEnd);
      }
    }
    return cnt;
  }
}

File: Sheets/src/neetcode/practice/Intervals/_3_NonOverlappingIntervals.cc
#include <algorithm>
#include <vector>

class Solution
{
public:
  int eraseOverlapIntervals(std::vector<std::vector<int>> &intervals)
  {
    auto INTERVAL_ORDER = [](const std::vector<int> &a,
                             const std::vector<int> &b) -> bool {
      return a[1] < b[1];
    };

    std::sort(intervals.begin(), intervals.end(), INTERVAL_ORDER);
    int n = static_cast<int>(intervals.size());

    int cnt{};
    int prevr{-50001};
    for (int i = 0; i < n; i++)
    {
      int l = intervals[i][0], r = intervals[i][1];
      if (l < prevr)
      {
        cnt += 1;
        continue;
      }
      prevr = r;
    }

    return cnt;
  }
};
